### DISCUSSION TITLE
Unable to match simulation results with Biot-Savart law

### DISCUSSION THREAD

**Discussion by conradtj1993** [magpylib/magpylib#601] [Q&A] [answered]
Hey, I am attempting to model a Helmholtz coil using Magpylib. As a sanity check, I modelled two coils of single windings. However, the result don't seem to match up. Is this expected? My thinking was that for a single wind, the simulation shouldn't actually be adding any insight because there are no (unique) approximations being made by the Biot-Savart analysis such as averaging the coil radii, so the results should be identical. Instead, the simulation yields a result approximately 20% lower than the analytical model. 

Perhaps I am wrong about this? I am wondering if this deviation is expected.

I've attached my code and a pdf detailing the differences in solutions. I apologize if this is below the normal level of rigor or the wrong style of question for this forum - its my first time posting on Github.

```
n = 1 #Number of winds
I = 1 #Current (A)
R  = 18 #Radius of winding (mm)
d = 23.2 #Coil separation (mm)

coil_1 =  magpy.Collection(style_label='coil_1')
coil_2 =  magpy.Collection(style_label='coil_2')

coil_1.add(Loop(current=I, diameter=2*R, position=(0,0,d/2)))
coil_2.add(Loop(current=I, diameter=2*R, position=(0,0,-d/2)))
helmholtz = coil_1 + coil_2

numObs = 500

Hz = np.zeros(numObs)
zMin = -20
z = np.arange(-zMin,zMin,(2*zMin)/numObs)
for ind in range(len(Hz)):
    Hz[ind] = helmholtz.getH((0,0,z[ind]))[2]
```

[MagpylibQuestion.pdf](https://github.com/magpylib/magpylib/files/11055611/MagpylibQuestion.pdf)

**Comment by OrtnerMichael** [comment, marked as answer]
hi @conradtj1993 

You question is well formatted, ty :) - so here is what i think:

----------------------------------------------
Are you sure this is the correct Biot-Savart solution ? You have two times the same term in the denominator. What you should get is two times this instead

![image](https://user-images.githubusercontent.com/47384012/227644601-5006cd53-dbb0-4a11-a908-53ef902dba6c.png)

this is from [hyperphysics](http://hyperphysics.phy-astr.gsu.edu/hbase/magnetic/curloo.html), and `magpylib.Loop` should give exactly this.

------------------------------------

Second thing: You have this line in your code

![image](https://user-images.githubusercontent.com/47384012/227644931-7ce9ee47-4e75-4591-a80f-5d9e83ecd858.png)

but you compute the field already with `getH`. So your field is already in kA/m. My guess is that you got the biot-savart right, but this is where your 20% come from.

-----------------------------

Third thing: 

In your code you compute the field at different positions using a loop.

```
Hz = np.zeros(numObs)
for ind in range(len(Hz)):
    Hz[ind] = helmholtz.getH((0,0,z[ind]))[2]
```

This is very inefficient, because the field computation function is called multiple times. Here one should make use of vectorization, i.e. provide the function with all positions at one time

```
pos = [(0,0,zz) for zz in z]
Hz2 = helmholtz.getH(pos)[:,2]
```

this gives the same result, but is at least 100 times faster. Vectorization is key to scientific programming with numpy, Matlab and a lot of other languages. Here is a reference

[Code Vectorization](https://en.wikipedia.org/wiki/Array_programming)

---------------------------------------------------------------


did this help ?

**Reply by conradtj1993** [reply]
Thank you for your thoughtful and comprehensive reply.

The equation you referenced is the correct solution for the field from a single loop. The superposition principle holds that the field generated by two loops is just the linear combination those produced by each loop individually, which is why my solution at the center point of the two coils looks like a sum of two of the hyperphyics solutions.

I have updated the code to implement vectorized inputs into the getH() function, this will be very helpful if I need to do any optimization work with my coils. Much appreciated.

The erroneous unit conversation was definitely the problem! Thank you, that was exactly it!
